File: mara_minimal_publisher/src/mara_minimal_publisher_v2.cpp
#include "rclcpp/rclcpp.hpp"
#include "hrim_actuator_rotaryservo_msgs/msg/goal_rotary_servo.hpp"
#include <chrono>

using namespace std::chrono_literals;

class MinimalPublisher : public rclcpp::Node
{
  public:
    MinimalPublisher() : Node("mara_minimal_publisher"),
    count_(0)
    {
      // Create a publisher on topic "/hrim_actuation_servomotor_000000000001/goal_axis1"
      pub_ = this->create_publisher<hrim_actuator_rotaryservo_msgs::msg::GoalRotaryServo>("/hrim_actuator_rotaryservo_000000000001/goal_axis1", rclcpp::SensorDataQoS());

      // Publishing rate of 1 Hz using a wall timer
      timer_ = this->create_wall_timer(1s, std::bind(&MinimalPublisher::timer_callback, this));
    }

  private:
    void timer_callback()
    {
      // Create message with the same type as the topic, <hrim_actuator_rotaryservo_msgs::msg::GoalRotaryServo>.
      auto msg = std::make_shared<hrim_actuator_rotaryservo_msgs::msg::GoalRotaryServo>();

      // Fill message content
      float position_deg = 30;
      msg->position = position_deg * 3.1416/180.0; // Position to rads
      msg->velocity = 0.4; // Velocity in rads/s
      msg->control_type = 4; // Position and velocity control

      // Publish message!
      pub_->publish(*msg);

      // Log
      RCLCPP_INFO(this->get_logger(), "Iteration number %d", count_);
      count_++;
    }

    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::Publisher<hrim_actuator_rotaryservo_msgs::msg::GoalRotaryServo>::SharedPtr pub_;
    size_t count_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);

  auto node = std::make_shared<MinimalPublisher>();
  rclcpp::spin(node);

  rclcpp::shutdown();
  return 0;
}

File: mara_minimal_publisher/scripts/mara_minimal_publisher_v2.py
#!/usr/bin/python3

import rclpy
from rclpy.node import Node
from rclpy.qos import qos_profile_sensor_data
from hrim_actuator_rotaryservo_msgs.msg import GoalRotaryServo

class MinimalPublisher(Node):

    def __init__(self):
        # Initialize Node with name "mara_minimal_publisher"
        super().__init__('mara_minimal_publisher')

        # Create a publisher on topic "/hrim_actuation_servomotor_000000000001/goal_axis1"
        self.pub_ = self.create_publisher(GoalRotaryServo, '/hrim_actuator_rotaryservo_000000000001/goal_axis1',
                                                qos_profile=qos_profile_sensor_data)

        # Create message with the same type as the topic, GoalRotaryServo
        self.msg = GoalRotaryServo()

        # Create a timer to publish the messages periodically
        timer_period = 1.0  # seconds
        self.timer = self.create_timer(timer_period, self.timer_callback)

        self.i = 0 # Iteration counter


    def timer_callback(self):
        # Fill message content
        position_deg = 30.
        self.msg.position = position_deg * 3.1416/180 # Position to rads
        self.msg.velocity = 0.4                       # Velocity in rads/s
        self.msg.control_type = 4                     # Position and velocity control

        # Publish message!
        self.pub_.publish(self.msg)

        # Log
        self.get_logger().info("Iteration number: {}".format(self.i))
        self.i += 1



def main(args=None):
    rclpy.init(args=args)

    minimal_publisher = MinimalPublisher()

    rclpy.spin(minimal_publisher)

    minimal_publisher.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()
